### ISSUE TITLE
ISO C structs with different tags are incompatible

### ISSUE THREAD

**Issue by Humm42** [cplusplus/draft#6791]
[diff.basic]/5 states:

> C allows “compatible types” in several places, C++ does not. For example, otherwise-identical ``struct`` types with different tag names are “compatible” in C but are distinctly different types in C++.

That hasn’t been true for C for quite a while now.

**Comment by frederick-vs-ja** [comment]
Looks like already false in C99 (perhaps changed by WG14 N522).

**Comment by frederick-vs-ja** [comment]
Perhaps we should say an enum type is compatible with its enum type in C, but not in C++.
```C++
int main(void) {
  enum et { v } e = v;
  unsigned int *p = e; // valid in C if the underlying type of enum et is unsigned int, invalid in C++
}
```

**Comment by eisenwave** [comment]
P2. If what you're saying is correct, the annex entry is factually incorrect and needs to be fixed.

I vaguely recall a paper or issue that pointed out we need to re-evaluate our [diff.iso] subclause for C23 in general. I cannot find it though.

**Comment by pmor13** [comment]
FYI:
1. The otherwise-identical struct types with different tag names are “compatible” in _ANSI_ C (see [Defect Report #029](https://www.open-std.org/jtc1/sc22/wg14/www/docs/dr_029.html)).
2. [N3032](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3032.pdf) proposes to "make tagged types with the same tag and content compatible inside the same translation unit".

@AaronBallman Does Clang [reject valid ANSI C code](https://godbolt.org/z/Pzre65an8)? In the [Defect Report #029](https://www.open-std.org/jtc1/sc22/wg14/www/docs/dr_029.html) the response was clear: "They (tags) are not used for type compatibility."

**Comment by AaronBallman** [comment]
> @AaronBallman Does Clang [reject valid ANSI C code](https://godbolt.org/z/Pzre65an8)? In the [Defect Report #029](https://www.open-std.org/jtc1/sc22/wg14/www/docs/dr_029.html) the response was clear: "They (tags) are not used for type compatibility."

Yes, intentionally (https://clang.llvm.org/c_dr_status.html) because of the danger of that kind of type confusion. We're not alone: https://godbolt.org/z/ha4oMb3vv (I am unaware of any C compiler which actually follows that DR; maybe one exists, but I think the market came to a different answer than the committee).

> [N3032](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3032.pdf) proposes to "make tagged types with the same tag and content compatible inside the same translation unit".

That was not adopted, but the follow-up [WG14 N3037](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3037.pdf) was adopted for C23.

**Comment by pmor13** [comment]
@AaronBallman Was "unaware of any C compiler which actually follows that DR" the rationale for the committee to use tags for type compatibility (in later revisions, e.g. in C11)?

**Comment by AaronBallman** [comment]
> [@AaronBallman](https://github.com/AaronBallman) Was "unaware of any C compiler which actually follows that DR" the rationale for the committee to use tags for type compatibility (in later revisions, e.g. in C11)?

C11 predates me joining WG14 (I joined during C17), so I'm not certain. As best I can tell, this was changed by https://www.open-std.org/jtc1/sc22/wg14/www/docs/n522.pdf during C99 but I've not dug out the committee minutes to see what the discussion was on the paper.
